File: erlang/robot-simulator/src/robot_simulator.erl
-module(robot_simulator).

-export([robot_loop/0, advance/1, create/0, direction/1, left/1, place/3, position/1, right/1]).

-type property() :: direction | position.
-type direction() :: north | east | south | west.
-type side() :: left | right.
-type position() :: {X::Integer, Y::Integer}.
-type robot() :: pid().

-spec move(direction(), position()) -> position().
move(Direction, {X,Y}) ->
    case Direction of
        north ->
            {X, Y+1};
        south ->
            {X, Y-1};
        east ->
            {X+1, Y};
        west ->
            {X-1, Y}
    end.

-spec turn(direction(), side()) -> direction().
turn(Direction, right) ->
    case Direction of
        north -> east;
        east -> south;
        south -> west;
        west -> north
    end;

turn(Direction, left) ->
    case Direction of
        north -> west;
        west -> south;
        south -> east;
        east -> north
    end.

robot_loop() ->
    receive
        {place, Direction, Position} -> 
            put(direction, Direction),
            put(position, Position);
        {advance} ->
            Direction = get(direction),
            Position = get(position),
            put(position, move(Direction, Position));
        {turn, Side} ->
            Direction = get(direction),
            put(direction, turn(Direction, Side));
        {From, direction} ->
            From ! {direction, get(direction)};
        {From, position} ->
            From ! {position, get(position)}
    end,
    robot_loop().

-spec create() -> robot().
create() ->
    Pid = spawn(fun robot_simulator:robot_loop/0),
    Pid ! {place, north, {0,0}},
    Pid.

-spec get_property(robot(), property()) -> direction() | position().
get_property(Robot, Prop) ->
    Robot ! {self(), Prop},
    receive
        {Prop, Value} -> Value
    end.

-spec direction(robot()) -> direction().
direction(Robot) -> get_property(Robot, direction).
-spec position(robot()) -> position().
position(Robot) -> get_property(Robot, position).

-spec advance(robot()) -> ok.
advance(Robot) -> 
    Robot ! {advance}, 
    ok.

-spec left(robot()) -> ok.
left(Robot) -> 
    Robot ! {turn, left}, 
    ok.

-spec right(robot()) -> ok.
right(Robot) -> 
    Robot ! {turn, right}, 
    ok.

-spec place(robot(), direction(), position()) -> ok.
place(Robot, Direction, Position) -> 
    Robot ! {place, Direction, Position},
    ok.

File: elixir/robot-simulator/robot_simulator.exs
defmodule RobotSimulator do
  @doc """
  Create a Robot Simulator given an initial direction and position.

  Valid directions are: `:north`, `:east`, `:south`, `:west`
  """
  @spec create(direction :: atom, position :: {integer, integer}) :: any
  def create(direction \\ nil, position \\ nil)
  def create(direction,_) when direction not in [:north, :east, :south, :west, nil], do: {:error, "invalid direction"}
  def create(direction, {x, y} = position) when is_number(x) and is_number(y) do
    out_direction = if nil == direction, do: :north, else: direction
    {out_direction, position}
  end
  def create(nil,nil), do: {:north, {0,0}}
  def create(_, _), do: {:error, "invalid position"}

  @doc """
  Simulate the robot's movement given a string of instructions.

  Valid instructions are: "R" (turn right), "L", (turn left), and "A" (advance)
  """
  @spec simulate(robot :: any, instructions :: String.t()) :: any
  def simulate(robot, instructions) do
    Enum.reduce_while(String.codepoints(instructions), robot,
      fn instruction, {direction, position} ->
        case turn(instruction, direction) do
          {:error, reason} ->
            {:halt, {:error, reason}}
          new_direction ->
            acc = {new_direction, advance(instruction, new_direction, position)}
            {:cont, acc}
        end
      end)
  end

  defp turn("R", direction) do
    case direction do
      :north -> :east
      :east -> :south
      :south -> :west
      :west -> :north
    end
  end

  defp turn("L", direction) do
    case direction do
      :north -> :west
      :west -> :south
      :south -> :east
      :east -> :north
    end
  end
  defp turn("A", direction), do: direction
  defp turn(_, _), do: {:error, "invalid instruction"}

  defp advance("A", :north, {x,y}), do: {x, y+1}
  defp advance("A", :west,  {x,y}), do: {x-1, y}
  defp advance("A", :south, {x,y}), do: {x, y-1}
  defp advance("A", :east, {x,y}), do: {x+1, y}
  defp advance(_, _, position), do: position

  @doc """
  Return the robot's direction.

  Valid directions are: `:north`, `:east`, `:south`, `:west`
  """
  @spec direction(robot :: any) :: atom
  def direction(robot) do
    elem(robot, 0)
  end

  @doc """
  Return the robot's position.
  """
  @spec position(robot :: any) :: {integer, integer}
  def position(robot) do
    elem(robot, 1)
  end
end
